client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -186,8 +205,33 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        synchronizer.updateState((state, updates) -> {
+            if (state.getConfigState() != ReaderGroupState.ConfigState.READY) {
+                val oldConfig = state.getConfig();
+                val newConfig = state.getNewConfig();
+                manageSubscriptions(oldConfig, newConfig);
+            }
+            updates.add(new ReaderGroupStateResetStart(config));
+        });
+
+        val generation = synchronizer.getState().getGeneration();
+        val currentConfig = synchronizer.getState().getConfig();
+        manageSubscriptions(currentConfig, config);
+
         Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        synchronizer.updateState((state, updates) -> {
+            updates.add(new ReaderGroupStateResetComplete(generation, config, segments, getEndSegmentsForStreams(config)));
+        });
+    }
+
+    private void manageSubscriptions(ReaderGroupConfig oldConfig, ReaderGroupConfig newConfig) {
+        Set<Stream> oldStreams = oldConfig.getRetentionConfig() != ReaderGroupConfig.RetentionConfig.NONE ? oldConfig.getStartingStreamCuts().keySet() : Collections.emptySet();
+        Set<Stream> newStreams = newConfig.getRetentionConfig() != ReaderGroupConfig.RetentionConfig.NONE ? newConfig.getStartingStreamCuts().keySet() : Collections.emptySet();
+        Set<String> streamsToSub = filterOut(newStreams, oldStreams).stream().map(Stream::getStreamName).collect(Collectors.toSet());
+        Set<String> streamsToUnsub = filterOut(oldStreams, newStreams).stream().map(Stream::getStreamName).collect(Collectors.toSet());
+        // subscriptions with the above gen
+        streamsToSub.forEach(s -> getThrowingException(controller.addSubscriber(scope, s, groupName)));
```

Comment:
Subscriber name passed to addSubscriber method should be 
= groupName + segmentNumber for its State Synchronizer Stream. Otherwise  when the same RG is deleted and recreated we'll have issues.
Same is the case for the `deleteSubscriber` and `updateRetentionStreamCut` calls.

Reply:
I've made the controller calls to accommodate both the segment number aspect and generations